File: sdk/typescript/src/engine.ts
/**
 * Execution engine. Walks a workflow tree and dispatches each node to a
 * caller-supplied HandlerResolver. Filter nodes are evaluated in-process.
 *
 * Two resolver strategies ship with the SDK:
 *   - InMemoryResolver: handlers are JS functions. Use in browser simulators.
 *   - HttpResolver:     handlers are remote workers. Use in Node deployments.
 */

import { evaluate } from "./condition.js";
import type {
  ExecutionResult,
  Node,
  NodeExecution,
  NodeResponse,
  Payload,
  Trace,
  WorkflowDef,
  WorkflowMessage,
} from "./types.js";

/** Resolves a node ID to an executor capable of running the node. */
export interface HandlerResolver {
  /**
   * Returns true iff the resolver can serve the given node ID.
   * Used at startup to verify workflow coverage.
   */
  has(nodeId: string): boolean;

  /** Execute a single Enhancer or Action node and return its response. */
  execute(msg: WorkflowMessage): Promise<NodeResponse>;
}

export interface EngineOptions {
  /** When true, fail any node whose output overwrites an input field. */
  strictMonotonicity?: boolean;
}

export class Engine {
  constructor(
    private readonly resolver: HandlerResolver,
    private readonly opts: EngineOptions = {},
  ) {}

  async execute(
    wf: WorkflowDef,
    executionId: string,
    input: Payload,
  ): Promise<ExecutionResult> {
    const nodes = new Map(wf.nodes.map((n) => [n.id, n]));
    const start = nodes.get(wf.start);
    if (!start) {
      throw new Error(`start node "${wf.start}" not found`);
    }

    const traces = await this.executeNode(wf, executionId, nodes, start, input, []);

    const allOK = traces.every((t) => t.status === "completed");
    return {
      execution_id: executionId,
      workflow_id: wf.id,
      status: allOK ? "completed" : "failed",
      traces,
    };
  }

  private async executeNode(
    wf: WorkflowDef,
    executionId: string,
    nodes: Map<string, Node>,
    node: Node,
    payload: Payload,
    parents: string[],
  ): Promise<Trace[]> {
    const exec: NodeExecution = {
      node_id: node.id,
      node_type: node.type,
      status: "running",
      input: { ...payload },
      started_at: new Date().toISOString(),
    };

    const msg: WorkflowMessage = {
      execution_id: executionId,
      workflow_id: wf.id,
      node_id: node.id,
      node_type: node.type,
      metadata: {
        workflow_version: wf.version ?? "",
        timestamp: new Date().toISOString(),
        parent_nodes: [...parents],
      },
      payload,
    };

    if (node.type === "filter") {
      return this.executeFilter(wf, executionId, nodes, node, msg, exec);
    }
    return this.executeLinear(wf, executionId, nodes, node, msg, exec);
  }

  private async executeLinear(
    wf: WorkflowDef,
    executionId: string,
    nodes: Map<string, Node>,
    node: Node,
    msg: WorkflowMessage,
    exec: NodeExecution,
  ): Promise<Trace[]> {
    if (!this.resolver.has(node.id)) {
      exec.status = "failed";
      exec.error = {
        code: "NO_WORKER",
        message: `no worker registered for node "${node.id}"`,
      };
      exec.ended_at = new Date().toISOString();
      return [{ nodes: [exec], status: "failed" }];
    }

    let resp: NodeResponse;
    try {
      resp = await this.resolver.execute(msg);
    } catch (err) {
      exec.status = "failed";
      exec.error = {
        code: "DISPATCH_ERROR",
        message: err instanceof Error ? err.message : String(err),
      };
      exec.ended_at = new Date().toISOString();
      return [{ nodes: [exec], status: "failed" }];
    }

    exec.ended_at = new Date().toISOString();

    if (resp.status === "failure") {
      exec.status = "failed";
      exec.output = resp.output;
      exec.error = resp.error;
      return [{ nodes: [exec], status: "failed" }];
    }

    exec.status = "completed";
    exec.output = resp.output;

    // Monotonicity check.
    const violations = checkMonotonicity(msg.payload, resp.output);
    if (violations.length > 0 && this.opts.strictMonotonicity) {
      exec.status = "failed";
      exec.error = {
        code: "MONOTONICITY_VIOLATION",
        message: `${node.type} node "${node.id}" overwrote input field(s): ${violations.join(", ")}`,
      };
      return [{ nodes: [exec], status: "failed" }];
    }

    const merged = { ...msg.payload, ...resp.output };

    if (!node.next_node) {
      return [{ nodes: [exec], status: "completed" }];
    }
    const next = nodes.get(node.next_node);
    if (!next) {
      exec.error = {
        code: "MISSING_NODE",
        message: `next node "${node.next_node}" not found`,
      };
      return [{ nodes: [exec], status: "failed" }];
    }

    const childTraces = await this.executeNode(
      wf,
      executionId,
      nodes,
      next,
      merged,
      [...msg.metadata.parent_nodes, node.id],
    );
    return childTraces.map((t) => ({
      nodes: [exec, ...t.nodes],
      status: t.status,
    }));
  }

  private async executeFilter(
    wf: WorkflowDef,
    executionId: string,
    nodes: Map<string, Node>,
    node: Node,
    msg: WorkflowMessage,
    exec: NodeExecution,
  ): Promise<Trace[]> {
    const hits: { label: string; next: Node }[] = [];

    for (let i = 0; i < (node.branches ?? []).length; i++) {
      const branch = node.branches![i]!;
      let activated: boolean;
      try {
        activated = evaluate(branch.condition, msg.payload);
      } catch (err) {
        exec.status = "failed";
        exec.error = {
          code: "CONDITION_ERROR",
          message: `branch ${i} (${branch.label ?? ""}): ${
            err instanceof Error ? err.message : String(err)
          }`,
        };
        exec.ended_at = new Date().toISOString();
        return [{ nodes: [exec], status: "failed" }];
      }
      if (!activated) continue;
      const next = nodes.get(branch.next_node);
      if (!next) {
        exec.status = "failed";
        exec.error = {
          code: "MISSING_NODE",
          message: `branch ${i} references undefined node "${branch.next_node}"`,
        };
        exec.ended_at = new Date().toISOString();
        return [{ nodes: [exec], status: "failed" }];
      }
      hits.push({ label: branch.label ?? next.id, next });
    }

    exec.ended_at = new Date().toISOString();
    exec.status = "completed";
    exec.output = { activated_branches: hits.map((h) => h.label) };

    if (hits.length === 0) {
      return [{ nodes: [exec], status: "completed" }];
    }

    // Execute every activated branch. Branches are independent (non-merging),
    // so we can run them in parallel via Promise.all.
    const childResults = await Promise.all(
      hits.map((h) =>
        this.executeNode(
          wf,
          executionId,
          nodes,
          h.next,
          msg.payload,
          [...msg.metadata.parent_nodes, node.id],
        ),
      ),
    );

    const allTraces: Trace[] = [];
    for (const traces of childResults) {
      for (const t of traces) {
        allTraces.push({
          nodes: [exec, ...t.nodes],
          status: t.status,
        });
      }
    }
    return allTraces;
  }
}

// --- Resolvers -------------------------------------------------------------

/** In-memory resolver: handlers are local async JS functions. */
export class InMemoryResolver implements HandlerResolver {
  private readonly handlers = new Map<
    string,
    (msg: WorkflowMessage) => Promise<NodeResponse> | NodeResponse
  >();

  register(
    nodeId: string,
    handler: (msg: WorkflowMessage) => Promise<NodeResponse> | NodeResponse,
  ): void {
    this.handlers.set(nodeId, handler);
  }

  has(nodeId: string): boolean {
    return this.handlers.has(nodeId);
  }

  async execute(msg: WorkflowMessage): Promise<NodeResponse> {
    const fn = this.handlers.get(msg.node_id);
    if (!fn) {
      throw new Error(`no handler for node "${msg.node_id}"`);
    }
    return await fn(msg);
  }
}

/**
 * HTTP resolver: dispatches each node to a remote worker.
 * Discovers handlers at construction time via GET /handlers.
 */
export class HttpResolver implements HandlerResolver {
  private readonly registry = new Map<string, string[]>();
  private readonly counter = new Map<string, number>();

  /** Construct a resolver and discover handlers from each worker URL. */
  static async discover(workerURLs: string[], timeoutMs = 5000): Promise<HttpResolver> {
    const r = new HttpResolver();
    for (const url of workerURLs) {
      const ids = await fetchHandlers(url, timeoutMs);
      r.register(url, ids);
    }
    return r;
  }

  register(workerURL: string, nodeIds: string[]): void {
    for (const id of nodeIds) {
      const list = this.registry.get(id) ?? [];
      if (!list.includes(workerURL)) list.push(workerURL);
      this.registry.set(id, list);
    }
  }

  has(nodeId: string): boolean {
    return (this.registry.get(nodeId)?.length ?? 0) > 0;
  }

  async execute(msg: WorkflowMessage): Promise<NodeResponse> {
    const urls = this.registry.get(msg.node_id) ?? [];
    if (urls.length === 0) {
      throw new Error(`no worker registered for node "${msg.node_id}"`);
    }
    const idx = (this.counter.get(msg.node_id) ?? 0) % urls.length;
    this.counter.set(msg.node_id, idx + 1);
    const url = urls[idx]!;

    const resp = await fetch(`${url}/execute`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(msg),
    });
    if (!resp.ok) {
      throw new Error(`worker ${url} returned ${resp.status}`);
    }
    return (await resp.json()) as NodeResponse;
  }
}

async function fetchHandlers(url: string, timeoutMs: number): Promise<string[]> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const r = await fetch(`${url}/handlers`, { signal: controller.signal });
    if (!r.ok) throw new Error(`worker ${url}/handlers returned ${r.status}`);
    const body = (await r.json()) as { node_ids?: string[] };
    return body.node_ids ?? [];
  } finally {
    clearTimeout(timer);
  }
}

function checkMonotonicity(input: Payload, output: Payload): string[] {
  const violations: string[] = [];
  for (const k of Object.keys(input)) {
    if (k in output && JSON.stringify(input[k]) !== JSON.stringify(output[k])) {
      violations.push(k);
    }
  }
  return violations;
}

File: internal/engine/engine.go
package engine

import (
	"fmt"
	"sync"
	"time"

	"github.com/harshaneel/steppr/internal/dispatch"
	"github.com/harshaneel/steppr/internal/model"
	"github.com/harshaneel/steppr/internal/registry"
)

type Engine struct {
	Dispatcher *dispatch.HTTPDispatcher
	Registry   *registry.Registry

	// StrictMonotonicity, when true, fails any node whose output overwrites a
	// field present in the input payload. When false (default), violations are
	// recorded as errors on the trace but execution continues.
	StrictMonotonicity bool
}

func New() *Engine {
	return &Engine{
		Dispatcher: dispatch.New(),
		Registry:   registry.New(),
	}
}

// resolveWorker returns the worker URL for a node, preferring the registry
// over a node-local Worker field. Empty string means no worker is available.
func (e *Engine) resolveWorker(node *model.Node) string {
	if e.Registry != nil {
		if url, ok := e.Registry.Resolve(node.ID); ok {
			return url
		}
	}
	return node.Worker
}

// checkMonotonicity returns the names of input fields that the worker's
// declared output also contains with a different value. An overwrite implies
// non-monotonic behavior.
func checkMonotonicity(input, output map[string]any) []string {
	if len(input) == 0 || len(output) == 0 {
		return nil
	}
	var violations []string
	for k, inV := range input {
		outV, ok := output[k]
		if !ok {
			continue
		}
		if !equalValues(inV, outV) {
			violations = append(violations, k)
		}
	}
	return violations
}

func equalValues(a, b any) bool {
	// Same-string comparison is a safe approximation for JSON-serializable
	// values. Slices, maps, etc. that differ in content will produce different
	// formatted strings.
	return fmt.Sprintf("%v", a) == fmt.Sprintf("%v", b)
}

func (e *Engine) Execute(wf *model.WorkflowDef, executionID string, input map[string]any) (*ExecutionResult, error) {
	nodes := wf.NodeMap()
	startNode, err := wf.StartNode()
	if err != nil {
		return nil, err
	}

	result := &ExecutionResult{
		ExecutionID: executionID,
		WorkflowID:  wf.ID,
		Status:      StatusRunning,
	}

	traces := e.executeNode(wf, executionID, nodes, startNode, input, nil)
	result.Traces = traces

	allOK := true
	for _, t := range traces {
		if t.Status == StatusFailed {
			allOK = false
			break
		}
	}
	if allOK {
		result.Status = StatusCompleted
	} else {
		result.Status = StatusFailed
	}

	return result, nil
}

func (e *Engine) executeNode(
	wf *model.WorkflowDef,
	executionID string,
	nodes map[string]*model.Node,
	node *model.Node,
	payload map[string]any,
	parentNodes []string,
) []Trace {
	msg := &model.WorkflowMessage{
		ExecutionID: executionID,
		WorkflowID:  wf.ID,
		NodeID:      node.ID,
		NodeType:    node.Type,
		Metadata: model.MessageMetadata{
			WorkflowVersion: wf.Version,
			Timestamp:       time.Now().UTC().Format(time.RFC3339),
			ParentNodes:     parentNodes,
		},
		Payload: payload,
	}

	exec := NodeExecution{
		NodeID:    node.ID,
		NodeType:  node.Type,
		Status:    StatusRunning,
		Input:     payload,
		StartedAt: time.Now(),
	}

	switch node.Type {
	case model.NodeTypeFilter:
		return e.executeFilter(wf, executionID, nodes, node, msg, exec)
	case model.NodeTypeEnhancer, model.NodeTypeAction:
		return e.executeLinear(wf, executionID, nodes, node, msg, exec)
	default:
		exec.Status = StatusFailed
		exec.Error = &model.ErrorInfo{Code: "INVALID_TYPE", Message: fmt.Sprintf("unknown node type: %s", node.Type)}
		exec.EndedAt = time.Now()
		return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusFailed}}
	}
}

func (e *Engine) executeLinear(
	wf *model.WorkflowDef,
	executionID string,
	nodes map[string]*model.Node,
	node *model.Node,
	msg *model.WorkflowMessage,
	exec NodeExecution,
) []Trace {
	workerURL := e.resolveWorker(node)
	if workerURL == "" {
		exec.EndedAt = time.Now()
		exec.Status = StatusFailed
		exec.Error = &model.ErrorInfo{
			Code:    "NO_WORKER",
			Message: fmt.Sprintf("no worker registered for node %q", node.ID),
		}
		return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusFailed}}
	}
	resp, err := e.Dispatcher.DispatchNode(workerURL, msg)
	exec.EndedAt = time.Now()

	if err != nil {
		exec.Status = StatusFailed
		exec.Error = &model.ErrorInfo{Code: "DISPATCH_ERROR", Message: err.Error()}
		return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusFailed}}
	}

	if resp.Status == "failure" {
		exec.Status = StatusFailed
		exec.Output = resp.Output
		exec.Error = resp.Error
		return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusFailed}}
	}

	exec.Status = StatusCompleted
	exec.Output = resp.Output

	// Monotonicity check: Enhancer and Action are required to preserve all input
	// fields (they may add but not remove). The orchestrator merges input into
	// output, so structural monotonicity is enforced at the data layer. We
	// additionally check whether the worker's *declared* output overwrites any
	// input fields, which would indicate the worker isn't honoring monotonicity.
	if violations := checkMonotonicity(msg.Payload, resp.Output); len(violations) > 0 {
		exec.Error = &model.ErrorInfo{
			Code:    "MONOTONICITY_VIOLATION",
			Message: fmt.Sprintf("%s node %q overwrote input field(s): %v", node.Type, node.ID, violations),
		}
		// Treat as warning-only by default: continue with merged payload but
		// record the violation. Strict mode could be enabled via env var.
		if e.StrictMonotonicity {
			exec.Status = StatusFailed
			return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusFailed}}
		}
	}

	// Merge output with input (monotonic: original data preserved, new data appended).
	merged := mergePayload(msg.Payload, resp.Output)

	// If terminal node, return single trace.
	if node.NextNode == "" {
		return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusCompleted}}
	}

	// Continue to next node.
	nextNode, ok := nodes[node.NextNode]
	if !ok {
		exec.Error = &model.ErrorInfo{Code: "MISSING_NODE", Message: fmt.Sprintf("next node %q not found", node.NextNode)}
		return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusFailed}}
	}

	childTraces := e.executeNode(wf, executionID, nodes, nextNode, merged, append(msg.Metadata.ParentNodes, node.ID))
	// Prepend this node's execution to all child traces.
	for i := range childTraces {
		childTraces[i].Nodes = append([]NodeExecution{exec}, childTraces[i].Nodes...)
	}
	return childTraces
}

// executeFilter evaluates each branch's condition against the current payload
// and dispatches to every branch whose condition holds. Filter is fully
// declarative: no worker call, no HTTP overhead. Multiple branches may
// activate simultaneously (OR-split).
func (e *Engine) executeFilter(
	wf *model.WorkflowDef,
	executionID string,
	nodes map[string]*model.Node,
	node *model.Node,
	msg *model.WorkflowMessage,
	exec NodeExecution,
) []Trace {
	type activated struct {
		label string
		next  *model.Node
	}
	var hits []activated

	for i, branch := range node.Branches {
		ok, err := branch.Condition.Evaluate(msg.Payload)
		if err != nil {
			exec.Status = StatusFailed
			exec.Error = &model.ErrorInfo{
				Code:    "CONDITION_ERROR",
				Message: fmt.Sprintf("branch %d (%s): %v", i, branch.Label, err),
			}
			exec.EndedAt = time.Now()
			return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusFailed}}
		}
		if !ok {
			continue
		}
		nextNode, found := nodes[branch.NextNode]
		if !found {
			exec.Status = StatusFailed
			exec.Error = &model.ErrorInfo{
				Code:    "MISSING_NODE",
				Message: fmt.Sprintf("branch %d references undefined node %q", i, branch.NextNode),
			}
			exec.EndedAt = time.Now()
			return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusFailed}}
		}
		hits = append(hits, activated{label: branch.Label, next: nextNode})
	}

	exec.EndedAt = time.Now()
	exec.Status = StatusCompleted
	labels := make([]string, 0, len(hits))
	for _, h := range hits {
		if h.label != "" {
			labels = append(labels, h.label)
		} else {
			labels = append(labels, h.next.ID)
		}
	}
	exec.Output = map[string]any{"activated_branches": labels}

	if len(hits) == 0 {
		// No branch activated — terminal trace.
		return []Trace{{Nodes: []NodeExecution{exec}, Status: StatusCompleted}}
	}

	// Execute each activated branch concurrently.
	var mu sync.Mutex
	var wg sync.WaitGroup
	var allTraces []Trace

	for _, h := range hits {
		wg.Add(1)
		go func(a activated) {
			defer wg.Done()
			childTraces := e.executeNode(wf, executionID, nodes, a.next, msg.Payload, append(msg.Metadata.ParentNodes, node.ID))
			mu.Lock()
			for i := range childTraces {
				childTraces[i].Nodes = append([]NodeExecution{exec}, childTraces[i].Nodes...)
			}
			allTraces = append(allTraces, childTraces...)
			mu.Unlock()
		}(h)
	}

	wg.Wait()
	return allTraces
}


func mergePayload(base, overlay map[string]any) map[string]any {
	merged := make(map[string]any, len(base)+len(overlay))
	for k, v := range base {
		merged[k] = v
	}
	for k, v := range overlay {
		merged[k] = v
	}
	return merged
}
